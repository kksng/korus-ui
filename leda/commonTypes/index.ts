import * as React from 'react';

export interface AttributesType {
  active?: boolean,
  box?: boolean,
  className?: string,
  contentBox?: boolean,
  dropdown?: boolean,
  menuBox?: string,
  navBox?: boolean,
  primary?: boolean,
  size?: string,
  userBox?: boolean,
  width?: number,
}

export interface ValidatedBlurEvent<T = HTMLInputElement> extends React.FocusEvent<T>{
  relatedTarget: EventTarget,
  target: EventTarget & T & { isValid?: boolean },
}

export interface DataObject { [x: string]: string | number }

export interface RenderEvent<P extends {} = {}, S extends {} = {}, E extends {} = {}> {
  Element: React.ElementType,
  componentProps: P,
  componentState: S,
  elementProps: E,
}

export interface CustomRender<P, S, E> {
  (props: RenderEvent<P, S, E>): React.ReactNode,
}

export interface SomeObject {
  [x: string]: any,
}

export interface SomeFunction {
  (x: unknown): unknown,
}

export interface SomeRefCurrent {
  [x: string]: Element | undefined,
}

export type RecursivePartial<T> = Partial<{
  [P in keyof T]?: T[P] extends object ? RecursivePartial<T[P]> : Partial<T[P]>;
}>;

export type RecursiveRequired<T> = Required<{
  [P in keyof T]?: T[P] extends object ? RecursiveRequired<T[P]> : Required<T[P]>;
}>;

export interface CustomEventHandler<T> {
  (ev: T): void,
}

export type Values<T> = T[keyof T];

export type SetState<S> = React.Dispatch<React.SetStateAction<S>>;

export interface Action<T, P = {}> {
  payload: P,
  type: T,
}

export type ArrayElement<A> = A extends readonly (infer T)[] ? T : A;

export interface CommonRefCurrent {
  wrapper: HTMLElement | null,
}
