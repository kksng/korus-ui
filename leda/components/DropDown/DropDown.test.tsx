import React from 'react';
import {
  render, screen, fireEvent,
} from '@testing-library/react';
import { DropDown } from './index';
// eslint-disable-next-line import/no-useless-path-segments
import * as L from '../../../leda';

describe('DropDown SNAPSHOTS', () => {
  it('should render', () => {
    const wrapper = render((
      <DropDown />
    ));

    expect(wrapper.container).toMatchSnapshot();
  });
});

describe('DropDown ATTRIBUTES', () => {
  describe('className prop', () => {
    it('should have className given by prop', () => {
      render((
        <DropDown _box>test</DropDown>
      ));

      expect(screen.queryByText('test')).toHaveClass('box');
    });

    it('should change classes through props', () => {
      render((
        <DropDown _active>test</DropDown>
      ));

      expect(screen.queryByText('test')).not.toHaveClass('box');

      expect(screen.queryByText('test')).toHaveClass('active');
    });

    it('className should not change prop-classes', () => {
      render((
        <DropDown _active className="test">test</DropDown>
      ));

      expect(screen.queryByText('test')).not.toHaveClass('box');

      expect(screen.queryByText('test')).toHaveClass('active');

      expect(screen.queryByText('test')).toHaveClass('test');
    });
  });

  it('should be wrapped in wrapper', () => {
    const wrapper = render((
      <DropDown />
    ));

    wrapper.rerender((
      <DropDown
        wrapperRender={() => (
          <span />
        )}
      />
    ));

    expect(document.querySelector('span')).toBeInTheDocument();

    wrapper.rerender((
      <DropDown
        wrapperRender={() => (
          <div />
        )}
      />
    ));

    expect(document.querySelector('div')).toBeInTheDocument();

    wrapper.rerender((
      <DropDown
        wrapperRender={() => (
          <p />
        )}
      />
    ));

    expect(document.querySelector('p')).toBeInTheDocument();
  });

  it('should have children prop', () => {
    render((
      <DropDown>
        <div className="level-1"><span className="level-2">test</span></div>
      </DropDown>
    )).debug();

    expect(screen.getByText('test').textContent).toEqual('test');

    expect(screen.getByText('test').tagName).toEqual('SPAN');

    expect(screen.getByText('test').className).toEqual('level-2');

    expect(screen.getByText('test').parentElement?.tagName).toEqual('DIV');

    expect(screen.getByText('test').parentElement?.className).toEqual('level-1');
  });

  describe('InteractionMode = "click"', () => {
    it('should be opened by click on and closed by click outside', () => {
      render((
        <DropDown interactionMode="click" wrapperRender={({ elementProps }: any) => <L.Button {...elementProps} />}>
          <L.Span>test</L.Span>
        </DropDown>
      ));

      fireEvent.click(screen.getByRole('button'));
      expect(screen.getByRole('button')).toHaveClass('opened');

      fireEvent.blur(screen.getByRole('button'));
      expect(screen.getByRole('button')).not.toHaveClass('opened');
    });
  });
});
