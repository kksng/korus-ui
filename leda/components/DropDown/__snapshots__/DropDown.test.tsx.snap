// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropDown SNAPSHOTS should render 1`] = `
<div>
  <div
    class="dropdown-wrapper"
  />
</div>
`;
